File: backend/app/Services/BusinessService.php
<?php

namespace App\Services;

use App\Models\Business;
use App\Models\Employee;
use App\Models\User;
use Illuminate\Support\Facades\DB;
use Exception;

class BusinessService
{
    public static function createBusiness(
        float $longitude,
        float $latitude,
        string $name,
        string $address,
        string $phone,
        string $type,
        User $user
    ): Business {
        return DB::transaction(function () use (
            $longitude,
            $latitude,
            $name,
            $address,
            $phone,
            $type,
            $user
        ) {
            // 1. Create Business
            $business = Business::create([
                'longitude' => $longitude,
                'latitude' => $latitude,
                'name' => $name,
                'address' => $address,
                'phone' => $phone,
                'type' => $type,
                'user_id' => $user->id,
            ]);

            // 2. Create Owner Role
            $roleOwner = EmployeeRoleService::createEmployeeRole(
                'Pemilik',
                'Pemilik UMKM',
                'Mengurus bisnis',
                true,
                $business->id
            );

            // 3. Add Owner as Employee
            $ownerEmployee = new Employee([
                'user_id' => $user->id,
                'role_id' => $roleOwner->id,
                'start_date' => now(),
                'salary' => 0.00,
            ]);
            EmployeeService::createEmployeeFromAdmin($user, $roleOwner, $ownerEmployee);

            return $business;
        });
    }

    public static function createBusinessWithEmployees(
        float $longitude,
        float $latitude,
        string $name,
        string $address,
        string $phone,
        string $type,
        User $user
    ): Business {
        return DB::transaction(function () use (
            $longitude,
            $latitude,
            $name,
            $address,
            $phone,
            $type,
            $user
        ) {
            // 1. Create Business + Owner Employee
            $business = self::createBusiness(
                $longitude,
                $latitude,
                $name,
                $address,
                $phone,
                $type,
                $user
            );

            // 2. Create Secondary User
            $employeeUser = UserService::registerUser(
                "employee{$business->id}",
                "employee{$business->id}@example.com",
                'password'

            );

            // 3. Create Employee Role for Secondary
            $roleEmployee = EmployeeRoleService::createEmployeeRole(
                'Pegawai',
                'Pegawai UMKM',
                'Mengurus bisnis',
                false,
                $business->id
            );

            // 4. Add Secondary Employee
            $secondaryEmployee = new Employee([
                'user_id' => $employeeUser->id,
                'role_id' => $roleEmployee->id,
                'start_date' => now(),
                'salary' => 0.00,
            ]);
            EmployeeService::createEmployeeFromAdmin($employeeUser, $roleEmployee, $secondaryEmployee);

            return $business;
        });
    }
}

File: kkn_backend/app/services/business_service.go
package services

import (
	"fmt"
	"kkn_backend/app/models"

	"github.com/goravel/framework/contracts/database/orm"
	"github.com/goravel/framework/facades"
)

func CreateBusiness(Longitude float64, Latitude float64, Name string, Address string, Phone string, Type models.BusinessType, user models.User) (*models.Business, error) {
	business := &models.Business{
		Longitude: Longitude,
		Latitude:  Latitude,
		Name:      Name,
		Address:   Address,
		Phone:     Phone,
		Type:      Type,
		UserId:    uint64(user.ID),
	}

	err := facades.Orm().Transaction(func(tx orm.Query) error {
		error := tx.Create(business)
		if error != nil {
			return error
		}
		businessId := business.ID
		// Create the owner then
		role, err := CreateRole(tx, "Pemilik", "Pemilik UMKM", "Mengurus bisnis", true, uint64(businessId))
		if err != nil {
			return err
		}
		// User owner as own employee (obviously duh)
		employee := &models.Employee{
			UserId: uint64(user.ID),
			RoleId: role.ID,
		}
		_, err = CreateEmployeeFromAdmin(&user, role, employee, tx)
		if err != nil {
			return err
		}
		// Create an employee for the CreateBusiness
		employeeUser, err := RegisterUser(tx, fmt.Sprintf("employee%d", businessId), fmt.Sprintf("employee%d@example.com", businessId), "password")
		if err != nil {
			return err
		}

		role2, err := CreateRole(tx, "Pegawai", "Pegawai UMKM", "Mengurus bisnis", false, uint64(business.ID))
		if err != nil {
			return err
		}
		employee2 := &models.Employee{
			UserId: uint64(employeeUser.ID),
			RoleId: uint64(role.ID),
		}
		_, err = CreateEmployeeFromAdmin(employeeUser, role2, employee2, tx)
		if err != nil {
			return err
		}
		if err := tx.Commit(); err != nil {
			return err
		}
		return nil
	})
	if err != nil {
		return nil, err
	}

	return business, nil
}
